presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1400,6 +1462,181 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
         }
     }
 
+    private static class PrestoS3StreamingOutputStream
+            extends OutputStream
+    {
+        private final AmazonS3 s3;
+        private final String bucketName;
+        private final String key;
+        private final String uploadId;
+
+        private final ExecutorService uploadExecutor;
+
+        private int currentPartNumber;
+        private byte[] buffer;
+        private int bufferSize;
+
+        private Future<UploadPartResult> inProgressUploadFuture;
+        private final List<UploadPartResult> parts = new ArrayList<>();
+
+        public PrestoS3StreamingOutputStream(
+                AmazonS3 s3,
+                String bucketName,
+                String key,
+                String uploadId,
+                ExecutorService uploadExecutor,
+                int partSize)
+        {
+            STATS.uploadStarted();
+
+            this.s3 = requireNonNull(s3, "s3 is null");
+
+            this.buffer = new byte[partSize];
+
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.uploadId = requireNonNull(uploadId, "uploadId is null");
+
+            this.uploadExecutor = requireNonNull(uploadExecutor, "uploadExecutor is null");
+        }
+
+        @Override
+        public void write(int b)
+                throws IOException
+        {
+            flushBuffer(false);
+            buffer[bufferSize] = (byte) b;
+            bufferSize++;
+        }
+
+        @Override
+        public void write(byte[] bytes, int offset, int length)
+                throws IOException
+        {
+            while (length > 0) {
+                int copied = min(buffer.length - bufferSize, length);
+                arraycopy(bytes, offset, buffer, bufferSize, copied);
+                bufferSize += copied;
+
+                flushBuffer(false);
```

Comment:
@chhsiao90 , @electrum : why would we want to try to flush buffer if buffer is not full?